memlimit/memlimit.go

```diff
@@ -85,9 +86,16 @@ func SetGoMemLimitWithProvider(provider Provider, ratio float64) (int64, error)
 	if err != nil {
 		return 0, err
 	}
-	goMemLimit := int64(ratio * float64(limit))
-	debug.SetMemoryLimit(goMemLimit)
-	return goMemLimit, nil
+
+	// If no hard memory limit is set, the provider will likely return something like math.MaxUint64,
+	// but that may not fit into an int64 used by debug.SetMemoryLimit().
+	if limit < math.MaxInt64 {
+		goMemLimit := int64(ratio * float64(limit))
+		debug.SetMemoryLimit(goMemLimit)
+		return goMemLimit, nil
+	} else {
+		return 0, errors.New("limit returned by provider is too high")
+	}
 }
 
```

Comment:
Thank you @LubosD! 

I investigated the spec about float to int conversion again, and I found that the result of conversion is implementation-dependent. It works differently depending on the arch. (FYI, https://github.com/golang/go/issues/45588)

Your suggestion looks good, but under the principle of least astonishment, I think it should also handle that limit exceeds the max value of the int64 without returning the error.

So, I would like to ask your opinion on the code below. If my suggestion seems reasonable to you, could you please modify the code according to my suggestion? If not, please feel free to comment.

```suggestion
	goMemLimit := cappedFloat2Int(float64(limit) * ratio)
	debug.SetMemoryLimit(goMemLimit)
	return goMemLimit, nil
}

func cappedFloat2Int(f float64) int64 {
	if f > math.MaxInt64 {
		return math.MaxInt64
	}
	return int64(f)
}
```